validator/sawtooth_validator/execution/scheduler_parallel.py

```diff
@@ -137,51 +184,51 @@ def find_write_predecessors(self, address):
         # Children readers must be added, since their reads must happen prior
         # to the write.
 
-        tokens = self._tokenize_address(address)
+        node_stream = self.walk(address)
 
         predecessors = set()
 
+        enclosing_writer = None
+
         # First, walk down from the root to the address, collecting all readers
         # and updating the enclosing_writer if needed.
 
-        node = self._root
-        enclosing_writer = node.writer  # possibly None
+        try:
+            for node in node_stream:
+                try:
+                    if node.writer is not None:
+                        enclosing_writer = node.writer
+                except AttributeError:
+                    if node is self.DUMMY:
+                        if enclosing_writer is not None:
+                            predecessors.add(enclosing_writer)
+                        break
+                    else:
+                        raise
 
-        # the readers at the root node will always be added
-        predecessors.update(set(node.readers))
+                # the readers at the root node will always be added
+                predecessors.update(set(node.readers))
+
+        except AddressNotInTree:
 
-        for token in tokens:
             # If the address isn't on the tree, then there aren't any
             # predecessors below the node to worry about (because
             # there isn't anything at all), so return the predecessors
             # that have already been collected.
-            if token not in node.children:
-                if enclosing_writer is not None:
-                    predecessors.add(enclosing_writer)
-                return predecessors
 
-            node = node.children[token]
+            if enclosing_writer is not None:
+                predecessors.add(enclosing_writer)
 
-            # add enclosing readers directly to predecessors
-            predecessors.update(set(node.readers))
+            return predecessors
 
-            if node.writer is not None:
-                enclosing_writer = node.writer
-
-        if enclosing_writer is not None:
-            predecessors.add(enclosing_writer)
+        # Next, descend down the tree starting at the address node and
+        # find all children writers and readers. Uses breadth first
+        # search.
```

Comment:
Please update the comments to be consistent. You mention earlier that the order is unspecified, but this says it is breadth first.